Better explanation of declaration vs. statement (#18724)
mdn/content: files/en-us/web/javascript/reference/statements/index.md

- {{jsxref("Statements/with", "with")}}
  - : Extends the scope chain for a statement.

## Difference between statements and declarations

In this section, we will be mixing two kinds of constructs: [_statements_](https://tc39.es/ecma262/#prod-Statement) and [_declarations_](https://tc39.es/ecma262/#prod-Declaration). They are two disjoint sets of grammars. The following are declarations:

- {{jsxref("Statements/let", "let")}}
- {{jsxref("Statements/const", "const")}}
- {{jsxref("Statements/function", "function")}}
- {{jsxref("Statements/function*", "function*")}}
- {{jsxref("Statements/async_function", "async function")}}
- {{jsxref("Statements/async_function*", "async function*")}}
- {{jsxref("Statements/class", "class")}}
- {{jsxref("Statements/export", "export")}} (Note: it can only appear at the top-level of a [module](/en-US/docs/Web/JavaScript/Guide/Modules))
- {{jsxref("Statements/import", "import")}} (Note: it can only appear at the top-level of a [module](/en-US/docs/Web/JavaScript/Guide/Modules))

Everything else in the [list above](#statements_and_declarations_by_category) is a statement.

> Statements and declarations are technical terms used in the formal syntax of JavaScript — colloquially, we would sometimes use them interchangeably. However, there's a noticeable difference in their usage. For example, in most control-flow structures, the body only accepts statements — such as the two arms of an [`if...else`](/en-US/docs/Web/JavaScript/Reference/Statements/if...else):

Reviewer: Minimally: "The terms "statement" and "declaration" are ...."

To my taste (entirely up to you):

```suggestion
The terms "statement" and "declaration" have a precise meaning in the formal syntax of JavaScript that affects where they may be placed in code.
For example, in most control-flow structures, the body only accepts statements — such as the two arms of an [`if...else`](/en-US/docs/Web/JavaScript/Reference/Statements/if...else):
```

```js
if (condition)
  statement1;
else
  statement2;
```

If you use a declaration instead of a statement, it would be a {{jsxref("SyntaxError")}}. For example, a [`let`](/en-US/docs/Web/JavaScript/Reference/Statements/let) declaration is not a statement, so you can't use it in its bare form as the body of an `if` statement.

```js example-bad
if (condition)
  let i = 0; // SyntaxError: Lexical declaration cannot appear in a single-statement context
```

On the other hand, [`var`](/en-US/docs/Web/JavaScript/Reference/Statements/var) is a statement, so you can use it on its own as the `if` body.

```js example-good
if (condition)

>   var i = 0;

Reviewer 2: Even `var i;` would be correct? (If I understand correctly)

Author: Yeah

```

You can see declarations as "binding identifiers to values", and statements as "carrying out actions". The fact that `var` is a statement instead of a declaration is a special case, because it doesn't follow normal lexical scoping rules and may create side-effects — in the form of creating global variables, mutating existing `var`-defined variables, and defining variables that are visible outside of its block.

As another example, [labels](/en-US/docs/Web/JavaScript/Reference/Statements/label) can only be attached to statements.

```js example-bad
label: const a = 1; // SyntaxError: Lexical declaration cannot appear in a single-statement context
```

> **Note:** there's a legacy grammar that allows [function declarations to have labels](/en-US/docs/Web/JavaScript/Reference/Statements/label#labeled_function_declarations), but it's only standardized for compatibility with web reality.

> You can always transform a declaration to a statement by wrapping it in braces — this makes it part of a [block statement](/en-US/docs/Web/JavaScript/Reference/Statements/block).

Reviewer 3: `const a = 1;` still remains a deceleration we are not changing it. we just want to wrap it in a block.

```suggestion
To get around this, you can wrap the declaration in braces — this makes it part of a [block statement](/en-US/docs/Web/JavaScript/Reference/Statements/block).
```
